collector/xml-eater.py

```diff
@@ -0,0 +1,48 @@
+from lxml import etree 
+import numpy as np
+import re
+#xml_file: file to be parsed
+#sample_rate: sample rate of audio that goes with xml_file
+#deliminator: deliminator for speaker labels.
+#             e.g. if we have "[M] Something something talking blah blah", 
+#                  then deliminator is a regex "\[[A-Z]\]". (match open square bracket, any single letter, close square bracket.)
+#discard: discard any lines that contain this regex (e.g., if they label sound effects in a consistant way, we can get rid of 
+#		those pretty easily using this param.)                      
+def xml_eater(xml_file, sample_rate, deliminator, discard=None):
+	# open xml
+	xml_file = open(xml_file).read()
+	tree = etree.fromstring(xml_file)	
+	starts = tree.xpath("text/@start")
+	texts  = [x.text.strip("\n") for x in tree.xpath("text")]
+	durs   = tree.xpath("text/@dur") 
+	speakers = [] # list of all possible speakers, used to determine size and entries of np array
+	# fill speakers
+	for text in texts:
+		for speaker in re.findall(deliminator, text):
+			if speaker not in speakers: speakers.append(speaker)
+	arr = np.zeros( (1, len(speakers)), dtype=np.int64)
+	current = None	
+	n = 0
+	# BUG: doesn't count time in between captions; TODO: fix this
+	while n < len(texts):
```

Comment:
If you want to hold onto the index while you loop, this is idiomatic:

```python
for n, test in enumerate(texts):
    ...
```

Reply:
I don't know where to comment for this, but when phrases extend across multiple captions, I (at least tried to) deal with that using the `current' variable. So current saves the speaker, and assigns the next speakerless texts to it, and then current changes when we have a new speaker.

Reply:
Yes, but say we find a speaker in row `i`, `current` gets a non-null value. Then we call check_valid(`i+1`) which returns False because row `i+1` has no speakers in it, so we assign `current = None`, but technically `i+1` was an extension of row `i`.

Reply:
Changed while loop into a for loop as you suggested.

I don't think that's right, if the text has no speakers in it, check_valid should return True:

if discard is not None and bool(re.search(discard, text)): return False
if len(re.findall(deliminator, text)) > 1: return False
if not bool(re.match(deliminator, text)) and re.search(deliminator, text): return False

first condition doesn't depend on whether there's no speakers, so we can ignore it.
if there's no speakers, re.findall(...) will return an empty list, so the second condition fails.
for the third condition, not re.match(...) will return True, but re.search(...) will return False (there's no speakers anywhere in the text, so it finds nothing), so the third if statement fails as well.

So we go to the default case of returning True.